### PR TITLE
Refactor GraphEdit connections

### PR DESCRIPTION
**Pull request by Jummit**
There was code to draw bezier curves in graph_edit.cpp, which I removed to use Curve2D instead.
This PR makes it possible to retrieve the points of a connection line between two given points, and to customize how connections are drawn:

<details>
  <summary>Examples of custom connections</summary>

```gdscript
func _get_connection_line(from: Vector2, to: Vector2):
	var distance = to - from
	return [from, from + Vector2(distance.x / 2, 0), to - Vector2(distance.x / 2, 0), to]
```

straight connection

```gdscript
func _get_connection_line(from: Vector2, to: Vector2):
	var distance = to - from
	return [from, to]
```

direct connection

When `_get_connection_line` isn't overriden:

curve

</details>  

By adding a way to access the connection line, it makes it possible to implement interaction, like putting nodes in between or adding reroute nodes:

demo

Gist:
https://gist.github.com/Jummit/545d193b9f5119a428862abfefe4be53

<details open>
  <summary>Issues to fix:</summary>

- [ ] not tested on the latest build, as my computer can't run Vulkan
- [x] Colors don't work <details><summary>Example</summary>
2020-07-20-122233_2646x1024_scrot
</details>

- [ ] I'm not sure about the get_connection_line name, as it's very close to get_connection_list.
- [ ] inconsistent with the old line when the connection goes backwards (or is this ok?)<details><summary>Example</summary>
image
Slide Comparison: https://imgsli.com/MTkyOTc
</details>
</details>

### CODE REVIEW

**Comment by swarnimarun** [comment]
I wanted to do it soo badly.
You have my thanks.
(Now if someone adds better styling support for graph/graphnodes and then zooming, I will be soo happy that I will scream... :p)

Once it's done I will see if I can find time to add these features to VisualScript and VisualShader Editor.

Will review it once I have some free time.

I have like 2 questions,
1. How does it handle overlapping?
2. What do you mean by wrong result when backwards.?
How is it affected by overloading get_connection_line?

Also you can do, "get_connection_spline" instead more technically accurate I think.

**Comment by Jummit** [comment]
> I wanted to do it soo badly.
> You have my thanks.

No problem.

>     1. How does it handle overlapping?

Like the old system, there should not be any changes.

>     2. What do you mean by wrong result when backwards.?
>        How is it affected by overloading get_connection_line?
 
It is inconsistent with the old look:

https://imgsli.com/MTkyOTc

> Also you can do, "get_connection_spline" instead more technically accurate I think.

It's only a spline if you use a Curve2D (the default), but you can add any line, as shown in the examples.

File: scene/gui/graph_edit.cpp [hunk 1]
```diff
@@ -599,63 +599,29 @@ bool GraphEdit::is_in_hot_zone(const Vector2 &pos, const Vector2 &p_mouse_pos) {
...
-	points.push_back(p_from);
 	colors.push_back(p_color);
-	_bake_segment2d(points, colors, 0, 1, p_from, c1, p_to, c2, 0, 3, 9, 3, p_color, p_to_color, lines);
-	points.push_back(p_to);
 	colors.push_back(p_to_color);
+	Vector<Vector2> points;
+	PackedVector2Array baked_points = get_connection_line(p_from, p_to);
+	for (int i = 0; i < baked_points.size(); i++) {
+		points.push_back(baked_points[i]);
+	}
```
Comment (Jummit) [inline scene/gui/graph_edit.cpp:624]: This is a manual conversion from `PackedVector2Array` to `Vector<Vector2>`, there must be a better solution.

**Comment by akien-mga** [comment]
Needs a rebase on current `master` to fix CI.

File: scene/gui/graph_edit.cpp (hunk 1 above)
Comment (swarnimarun) [reply scene/gui/graph_edit.cpp:624]: but why?
https://github.com/godotengine/godot/blob/d14932aecae5f380d766d83a8f292dd1dbf97540/core/variant.h#L66

**Comment by swarnimarun** [comment]
> > Also you can do, "get_connection_spline" instead more technically accurate I think.
> 
> It's only a spline if you use a Curve2D (the default), but you can add any line, as shown in the examples.

I mean splines can be linear as well, **spline represents a set of line segments between the two adjacent data points**

File: scene/gui/graph_edit.cpp (hunk 1 above)
Comment (Jummit) [reply scene/gui/graph_edit.cpp:624]: Casting directly from Vector<Vector2> to PackedVector2Array is not possible in 3.2, but seems to work in 4.0.

**Comment by Jummit** [comment]
> Needs a rebase on current `master` to fix CI.

Done.

**Comment by fire** [comment]
I and @swarnimarun will review on a spare moment.

**Comment by fire** [comment]
I'm looking at the code and the resulting lines are too thin.

image


It seems to be related to

> p_where->draw_polyline_colors(points, colors, 2 * EDSCALE);

However, I don't remember where the default for this is.

**Comment by fire** [comment]
RenderingServerCanvas::canvas_item_add_polyline has not implemented width. I'll have to circle back to that.

**Comment by aaronfranke** [comment]
@Jummit Is this still desired? If so, it needs to be [rebased](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) and tested on the latest master branch.

**Comment by Jummit** [comment]
I'm hesitant to work on it because I can't test the master branch, and there are some graphical inconsistencies I would need to look into. So closing this, but it's pretty salvageable if anyone would like to pick it up. If I ever get a PC that supports Vulkan I'll rebase and test it.

**Pull request closed without merging**